### PR TITLE
Rewrite of nfs.c to keep options per host separated.

### PR DESCRIPTION
**Pull request by FransUrbo**
Example: Setting sharenfs to
```
  "rw=@192.168.69.8,no_subtree_check,no_root_squash,ro=@192.168.69.45,no_subtree_check,root_squash"
```
will create the option string
```
  sec=sys,ro,no_subtree_check,no_root_squash,mountpoint,no_subtree_check,no_root_squash,no_subtree_check,root_squash
```
which is then run on _each host_!. Part of this is "default" options.

Instead, go through the option string, look for hosts and then make sure options for each host is kept apart by using a linked list (maintained by the ZFS list_* functions).
Also, the default options should only be added to the list of options if setting sharenfs to <code>on</code>.

Closes: #2773

### CODE REVIEW

**Comment by FransUrbo** [comment]
#3738

**Comment by FransUrbo** [comment]
@behlendorf Do you think you can spend an hour or two to look this over? It's now been running in the Debian GNU/Linux packages for almost a year and except for a minor problem in the beginning, I've heard nothing about this so I'm pretty sure it works just fine.

File: lib/libshare/nfs.c [hunk 1]
```diff
@@ -29,12 +29,18 @@
...
 #include <sys/wait.h>
+#include <sys/list.h>
 #include <unistd.h>
 #include <libzfs.h>
 #include <libshare.h>
 #include "libshare_impl.h"
 
+#if !defined(offsetof)
+#define	offsetof(s, m)  ((size_t)(&(((s *)0)->m)))
+#endif
```
Comment (behlendorf) [inline lib/libshare/nfs.c]: Why was this needed, `offsetof` is part of stddef.h
http://man7.org/linux/man-pages/man3/offsetof.3.html

File: lib/libshare/nfs.c [hunk 2]
```diff
@@ -44,93 +50,37 @@ static sa_fstype_t *nfs_fstype;
...
-			if (rc != SA_OK) {
-				free(host_dup);
-
-				return (rc);
-			}
+/* List of hosts and their options */
+/* NOTE: share path is stored elsewhere */
+typedef struct nfs_share_list_s {
+	char	host[255];
+	char	opts[255];
```
Comment (behlendorf) [inline lib/libshare/nfs.c]: Let's add #define's for these buffer sizes.  It may make sense to move those defines and this structure to `lib/libshare/nfs.h` which would be analogous to `lib/libshare/smb.h`.

File: lib/libshare/nfs.c [hunk 3]
```diff
@@ -149,13 +99,11 @@ get_linux_hostspec(const char *solaris_hostspec, char **plinux_hostspec)
 		 * to skip the @ in this case
 		 */
 		*plinux_hostspec = strdup(solaris_hostspec + 1);
-	} else {
+	} else
 		*plinux_hostspec = strdup(solaris_hostspec);
-	}
```
Comment (behlendorf) [inline lib/libshare/nfs.c]: [cstyle] The { } are expected for the `else` block when need by the `if` portion.

File: lib/libshare/nfs.c [hunk 4]
```diff
@@ -164,45 +112,23 @@ get_linux_hostspec(const char *solaris_hostspec, char **plinux_hostspec)
...
+	hostpath = malloc(strlen(host) + 1 +
+	    strlen(sharepath) + 1);
+	if (hostpath == NULL)
+		return (SA_NO_MEMORY);
+	sprintf(hostpath, "%s:%s", host, sharepath);
 
 #ifdef DEBUG
-	fprintf(stderr, "sharing %s with opts %s\n", hostpath, opts);
+	fprintf(stderr, "sharing %s with opts %s\n",
+		hostpath, (opts ? opts : "NULL"));
```
Comment (behlendorf) [inline lib/libshare/nfs.c]: [cstyle] 4 space indent to continue a line.

File: lib/libshare/nfs.c [hunk 5]
```diff
@@ -212,10 +138,17 @@ nfs_enable_share_one(const char *sharepath, const char *host,
 	argv[4] = hostpath;
 	argv[5] = NULL;
 
+#if DEBUG >= 2
+	int i;
+	fprintf(stderr, "CMD: ");
+	for (i = 0; i < 5; i++)
+		fprintf(stderr, "%s ", argv[i]);
+	fprintf(stderr, "\n");
+#endif
```
Comment (behlendorf) [inline lib/libshare/nfs.c]: Do we still need these DEBUG blocks.  If we really need to preserve this debugging I suggest updating the code to use `dprintf()`.  Then this debugging can be enabled at run time rather than build time like this, `ZFS_DEBUG="on" zfs share ...`.

File: lib/libshare/nfs.c [hunk 6]
```diff
@@ -258,19 +195,104 @@ add_linux_shareopt(char **plinux_opts, const char *key, const char *value)
...
+ * in "cookie"..
  */
 static int
 get_linux_shareopts_cb(const char *key, const char *value, void *cookie)
 {
 	char **plinux_opts = (char **)cookie;
+	nfs_share_list_t *opts = NULL;
+
+// TODO: Make sure that a 'sec=yyy' (or any other option) BEFORE a/any
+//       'rw[=@xxx]' or 'ro[=@xxx]' works.
```
Comment (behlendorf) [inline lib/libshare/nfs.c]: Please verify that this does work and then remove to the TODO comment.

File: lib/libshare/nfs.c [hunk 7]
```diff
@@ -258,19 +195,104 @@ add_linux_shareopt(char **plinux_opts, const char *key, const char *value)
...
+	if (strcmp(key, "ro") == 0 || strcmp(key, "rw") == 0) {
+		opts = (nfs_share_list_t *)
+			malloc(sizeof (nfs_share_list_t));
```
Comment (behlendorf) [inline lib/libshare/nfs.c]: [cstyle] 4 space indent.  Although in this case it should all fit on one line.

File: lib/libshare/nfs.c [hunk 8]
```diff
@@ -320,21 +355,36 @@ get_linux_shareopts(const char *shareopts, char **plinux_opts)
...
+		(void) add_linux_shareopt(plinux_opts, "no_subtree_check",
+		    NULL);
+		(void) add_linux_shareopt(plinux_opts, "no_root_squash", NULL);
+		(void) add_linux_shareopt(plinux_opts, "mountpoint", NULL);
+// TODO: Make sure a 'sharenfs=zzz' WITHOUT a 'rw[=@xxx]' or 'ro[=@xxx]' works
+//	} else if(strcmp(shareopts, "ro\0") != 0) {
+//		/* 'else' only: Doesn't work if there's a 'ro[=@xxx]' somewhere. */
+//		/* 'else if(strcmp(shareopts, "ro\0") == 0)': Doesn't work if there's ONLY a 'ro'. */
+//fprintf(stderr, "  + adding 'rw'\n"); // DEBUG
+//		strcat((char *)shareopts, ",rw");
```
Comment (behlendorf) [inline lib/libshare/nfs.c]: Does this work as expected?  If so remove the TODO.  If not it should be fixed.

File: lib/libshare/nfs.c [hunk 9]
```diff
@@ -344,24 +394,30 @@ get_linux_shareopts(const char *shareopts, char **plinux_opts)
...
 		return (SA_SYSTEM_ERR);
-	}
 
 	shareopts = FSINFO(impl_share, nfs_fstype)->shareopts;
-
 	if (shareopts == NULL)
 		return (SA_OK);
 
-	rc = get_linux_shareopts(shareopts, &linux_opts);
+	linux_opts = calloc(sizeof (nfs_share_list_t), 1);
```
Comment (behlendorf) [inline lib/libshare/nfs.c]: This needs to be checked for a NULL return.

File: lib/libshare/nfs.c [hunk 10]
```diff
@@ -404,6 +450,14 @@ nfs_disable_share_one(const char *sharepath, const char *host,
 	argv[2] = hostpath;
 	argv[3] = NULL;
 
+#if DEBUG >= 2
+	int i;
+	fprintf(stderr, "CMD: ");
+	for (i = 0; i < 3; i++)
+		fprintf(stderr, "%s ", argv[i]);
+	fprintf(stderr, "\n");
+#endif
```
Comment (behlendorf) [inline lib/libshare/nfs.c]: Drop or convert to dprintf()

File: lib/libshare/nfs.c [hunk 11]
```diff
@@ -420,6 +474,14 @@ nfs_disable_share_one(const char *sharepath, const char *host,
 static int
 nfs_disable_share(sa_share_impl_t impl_share)
 {
+	int rc = SA_OK;
+	char *shareopts, *linux_opts;
+	nfs_share_list_t *share;
+
+#ifdef DEBUG
+	fprintf(stderr, "nfs_disable_share: %s\n", impl_share->sharepath);
+#endif
```
Comment (behlendorf) [inline lib/libshare/nfs.c]: Drop or convert to dprintf()

File: lib/libshare/nfs.c [hunk 12]
```diff
@@ -428,7 +490,27 @@ nfs_disable_share(sa_share_impl_t impl_share)
 		return (SA_OK);
 	}
 
-	return (foreach_nfs_host(impl_share, nfs_disable_share_one, NULL));
+	shareopts = FSINFO(impl_share, nfs_fstype)->shareopts;
+	if (shareopts == NULL)
+		return (SA_OK);
+
+	linux_opts = calloc(sizeof (nfs_share_list_t), 1);
```
Comment (behlendorf) [inline lib/libshare/nfs.c]: This needs to be checked for a NULL return.

File: lib/libshare/nfs.c [hunk 13]
```diff
@@ -440,6 +522,8 @@ nfs_validate_shareopts(const char *shareopts)
 	char *linux_opts;
 	int rc;
 
+	linux_opts = calloc(sizeof (nfs_share_list_t), 1);
```
Comment (behlendorf) [inline lib/libshare/nfs.c]: This needs to be checked for a NULL return.

**Comment by behlendorf** [comment]
Closing for now to reduce the number of open active PRs.  We can absolutely consider merging this functionality once the patch is refreshed against master and we have some share test cases to validate this works.

File: lib/libshare/nfs.c (hunk 2 above)
Comment (FransUrbo) [inline lib/libshare/nfs.c]: This is a NFS limitation, so I don't see the point in having this as a define..

https://linux.die.net/man/5/nfs
```
namlen=n

The maximum length of a pathname component on this mount. If this option is not specified, the maximum length is negotiated with the server. In most cases, this maximum length is 255 characters.
```

**Comment by dipack95** [comment]
@behlendorf @FransUrbo Issue #2773 still plagues us, and I noticed this patch attempts to fix that issue. Is there an ETA on this being merged in the project? I can help if required.
